rfcs/2020-06-18-2625-architecture-revisit.md

```diff
@@ -0,0 +1,504 @@
+
+1. We are adding both implementation complexity as well as increasing the size
+   of the mental model needed to understand Vector's behavior.
+
+2. It's possible that tying computational concurrency to source task concurrency
+   results will work well in some situations and work poorly in others. It adds
+   a significant level of variability to something that used to be relatively
+   consistent. This applies to both performance as well as memory use.
+
+3. Many of the benefits here will not come automatically. We'll need to make
+   sure that we keep the overall design in mind and make individual components
+   that work well with it. This applies particularly to sources that don't
+   currently run more than one task (e.g. file source, UDP-based sources, etc).
+
+4. It's not entirely clear how runtimes like Lua and WASM should fit into this
+   model. Their biggest benefit is the flexibility to do anything, which goes
+   against our strategy of limiting transform capability. They also have
+   relatively unknown memory requirements, which could make them unsuitable to
+   run potentially thousands of copies of in a high-concurrency server scenario.
+   The best solution here may be something like the Lua transform's hooks, which
+   we can inspect at runtime to determine the required capabilities.
+
+## Rationale and Alternatives
+
+The rationale for these changes is to give us the maximum amount of future
+flexibility for the least amount of work. If we can isolate as much code as
+possible from the effects of future changes, we should be in a good position to
+move forward smoothly.
+
+The biggest alternative to doing this kind of reworking would be to stick with
+the existing overall model and try to work things like concurrency into each
+individual component. One example proposal along these lines was to have
+transforms batch up events and spawn a new task to process them. This would
+likely require some degree of similar changes in terms of defining types of
+transforms, but would keep topology-level architecture untouched.
+
+Making fewer, less invasive changes would have the strong benefit of being
+cheaper, but I worry that the improvements we'd be able to make with such
+a limited view (i.e. only within a single component) would keep us rather
+limited.
+
+When considering specifically the change to push transforms into source tasks,
+there are a number of alternatives for getting that kind of concurrency. While
+I believe that design has the most promise, one benefit of this proposal is that
+it becomes an implementation detail. The other changes around component and
+topology design should give us the freedom to experiment with different ways of
+achieving concurrency under the hood.
+
+## Outstanding Questions
+
+1. Where do runtimes fit into the new model of transforms?
+
+2. Should we (and if so, how should we) add concurrency to components like the
+   file source and UDP-based socket and syslog?
```

Comment:
For the `file` source I vote for one task per file. I don't know how else you'd draw lines there. I realize this could cause problems for rapidly rotating files, but that is not the common use case we are solving for.

For UDP-based sources, I'm curious if we could scale out based on the `host` value. I'm curious what @bruceg thinks.

Reply:
It would be possible but has some complications. The biggest I am aware of is that it would require manually handing off packets from a single receiver task to sub-tasks for each host. For some sources this context switching could significantly reduce the benefits of the scaling, although it would gain the benefits of separate threads for the transforms. This setup would also have to manually handle timing out the sub-tasks when senders go idle.

Reply:
Yeah, there are a few options here. We could build it directly into the sink as @bruceg said, which is a little annoying but theoretically not too complicated. Another option (which I think I lean towards), is adding some idea of explicit partitioning to the pipeline API and then using that. It'd be more work to design, but the sources would be simpler and we'd be able to reuse it for configurable partitioning in the future.